### PR TITLE
[WIP] Docs Refactor

### PR DESCRIPTION
**Pull request by trhr**
#### Brief overview of PR changes/additions
- Reorganize docs into folders
- Support extended functionality

#### Motivation for adding to evennia
- I needed to read the docs

### CODE REVIEW

**Comment by trhr** [comment]
Sorry for the messy branch. Github workflow stuff shouldn't have come over with this.

**Comment by trhr** [comment]
Still to do:

- Support linking straight into API
- Swap [text](url#inner) to [text](url:inner) format
- Other stuff

File: .github/workflows/pythonpackage.yml [hunk 1]
```diff
@@ -0,0 +1,124 @@
+# This workflow will install Python dependencies, run tests and lint with a variety of Python versions
```
Comment (Griatch) [inline .github/workflows/pythonpackage.yml:1]: This seems unrelated to the scope of this PR. If you want to suggest using Github workflows over Travis, please make a separate PR.

File: .travis/mysql_settings.py [hunk 2]
```diff
@@ -44,8 +44,8 @@
         "NAME": "evennia",
         "USER": "evennia",
         "PASSWORD": "password",
-        "HOST": "localhost",
-        "PORT": "",  # use default port
+        "HOST": "127.0.0.1",
+        "PORT": "3306",  # use default port
```
Comment (Griatch) [inline .travis/mysql_settings.py:48]: Comments would be good. Is there a purpose to these changes?

File: docs/broken_links.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import re
```
Comment (Griatch) [inline docs/broken_links.py:1]: This utility script appears to belong under `docs/pylib/`

Comment (Griatch) [reply docs/broken_links.py:1]: It also requires a doc string, like every other python resource.

File: docs/broken_links.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+parser.add_argument('--verbose', default=False, help="Verbosity level (1: Show all links)")
+
+args = parser.parse_args()
+
+def shortpath(path):
+    return os.path.relpath(path)
+
+def check_broken_links(file):
+    global broken_count
+    global clean_count
```
Comment (Griatch) [inline docs/broken_links.py:23]: Meaningless to use `global` here, when these things can easily be returned from the function.

File: docs/broken_links.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                    else:
+                        if args.verbose:
+                            print(Fore.GREEN, "\t\tClean : {link}{extra}".format(link=shortpath(link_dest), extra=extra), Style.RESET_ALL)
+                            clean_count += 1
+        else:
+            if args.verbose:
+                print(Fore.RED, "{file} has no links".format(file=shortpath(file)), Style.RESET_ALL)
+
+def get_md_files(path):
+    global files
```
Comment (Griatch) [inline docs/broken_links.py:44]: Same thing here, just pass this into the function.

File: docs/broken_links.py (hunk 3 above)
Comment (Griatch) [reply docs/broken_links.py:1]: A broken-link checker is a useful resource though! :+1:

File: docs/source/config.json [hunk 6]
```diff
@@ -0,0 +1,12 @@
+{
+    "replacementPatterns": [
+        {
+            "pattern": ".md$",
+            "replacement": ""
+        },
+	{
+            "pattern": "$",
+            "replacement": ".md"
+	}
```
Comment (Griatch) [inline docs/source/config.json:10]: What does this do?

Comment (trhr) [reply docs/source/config.json:10]: Nothing, file should be removed. Was part of the broken link checker.

File: docs/broken_links.py (hunk 5 above)
Comment (trhr) [reply docs/broken_links.py:44]: Not part of the docs refactor, just a tool I made for my own workflow.

File: docs/broken_links.py (hunk 3 above)
Comment (trhr) [reply docs/broken_links.py:1]: Not part of the docs refactor, just a tool I made for my own workflow.

File: .travis/mysql_settings.py (hunk 2 above)
Comment (trhr) [reply .travis/mysql_settings.py:48]: Not part of the docs refactor.

**Comment by trhr** [comment]
My recommendation for reviewing would be to pull and review the documentation as a whole rather than line by line. It's not code. We basically need to determine things like:

- Does it look good
- Does the new user workflow function
- Is it easy for experienced developers to find specific things

Those will never appear in a line by line.

**Comment by trhr** [comment]
As far as the quantitative text changes, yes, they are substantial as well.

- Entire sections were removed as redundant, interlinked, or split into separate pages
- The entire "getting started" workflow has been rejiggered
- Virtualenv documentation was from python2, updated to python3
- Less technical pages like 'contributing' often got a rewrite/restructure
- Links to Django 1.6 were updated to Django 3.0
- Links to the wiki were replaced with internal links

Unless I saw a specific change that needed to happen in a page in the evennia_core or tutorials_and_examples directories, those are almost entirely unchanged for content. The rest of the directories, as well as the 'hub' type pages, are fair game for sweeping changes.

I wish I had known you wanted to line-by-line this before I started, I expected with it being 'not ready for this scale of work' the whole point was that I bring it up to the point where it is ready for this scale. 

It is ready for a structural review, but I expect the full internal review will take months after the docs are live.

**Comment by Griatch** [comment]
@trhr The problem with sweeping changes like this is that if there are issues with the change, it's very hard to find those among changes that just move stuff around. A line-by-line is not needed, but it would still help if you could self-review and make github comments to point out your thinking and reasoning for changes while you still have this fresh in your mind.

**Comment by Griatch** [comment]
@trhr ... But I should also not forget to thank you for your effort on doing this; it's a lot of work spent here! :D

File: docs/source/evennia_core/system/commands/Command-Sets.md [hunk 7]
```diff
@@ -5,13 +5,13 @@ Command Sets are intimately linked with [Commands](Commands) and you should be f
```
Comment (trhr) [inline docs/source/evennia_core/system/commands/Command-Sets.md:5]: no changes

File: docs/source/evennia_core/system/commands/Command-System.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# Command System
```
Comment (trhr) [inline docs/source/evennia_core/system/commands/Command-System.md:1]: no changes

File: docs/source/evennia_core/system/django/serverconf.md [hunk 9]
```diff
@@ -0,0 +1,4 @@
+## ServerConf
```
Comment (trhr) [inline docs/source/evennia_core/system/django/serverconf.md:1]: this is deep deep stuff, probably should be its own page and not ever really mentioned in the rest of the docs. people can find it in the toctree if they want it.

File: docs/source/evennia_core/system/help/Help-System.md [hunk 10]
```diff
@@ -47,7 +47,7 @@ The text at the very top of the command class definition is the class' `__doc__`
```
Comment (trhr) [inline docs/source/evennia_core/system/help/Help-System.md:47]: no changes

File: docs/source/evennia_core/system/locks/Locks.md [hunk 11]
```diff
@@ -0,0 +1,197 @@
+```python
```
Comment (trhr) [inline docs/source/evennia_core/system/locks/Locks.md:1]: no changes

File: docs/source/evennia_core/system/locks/permissions.md [hunk 12]
```diff
@@ -0,0 +1,168 @@
+# Permissions
```
Comment (trhr) [inline docs/source/evennia_core/system/locks/permissions.md:1]: no changes

File: docs/source/evennia_core/system/objects/Character.md [hunk 13]
```diff
@@ -0,0 +1,11 @@
+# Characters
```
Comment (trhr) [inline docs/source/evennia_core/system/objects/Character.md:1]: no changes

File: docs/source/evennia_core/system/objects/Exit.md [hunk 14]
```diff
@@ -0,0 +1,28 @@
+# Exits
```
Comment (trhr) [inline docs/source/evennia_core/system/objects/Exit.md:1]: no changes

File: docs/source/evennia_core/system/objects/Objects.md [hunk 15]
```diff
@@ -0,0 +1,82 @@
+# Objects
```
Comment (trhr) [inline docs/source/evennia_core/system/objects/Objects.md:1]: no changes

File: docs/source/evennia_core/system/objects/Room.md [hunk 16]
```diff
@@ -0,0 +1,3 @@
+# Rooms
```
Comment (trhr) [inline docs/source/evennia_core/system/objects/Room.md:1]: no changes

File: docs/source/evennia_core/system/portal/Custom-Protocols.md [hunk 17]
```diff
@@ -5,8 +5,8 @@
 their own custom client protocol.*
```
Comment (trhr) [inline docs/source/evennia_core/system/portal/Custom-Protocols.md:5]: no changes

File: docs/source/evennia_core/system/portal/Inputfuncs.md [hunk 18]
```diff
@@ -31,7 +31,7 @@ Evennia defines a few default inputfuncs to handle the common cases. These are d
  - Input: `("text", (textstring,), {})`
```
Comment (trhr) [inline docs/source/evennia_core/system/portal/Inputfuncs.md:31]: no changes

File: docs/source/evennia_core/system/portal/Messagepath.md [hunk 19]
```diff
@@ -22,7 +22,7 @@ The client sends data to Evennia in two ways.
  - When first connecting, the client can send data to the server about its
```
Comment (trhr) [inline docs/source/evennia_core/system/portal/Messagepath.md:22]: no changes

File: docs/source/evennia_core/system/portal/OOB.md [hunk 20]
```diff
@@ -8,7 +8,7 @@ Inside Evennia, all server-client communication happens in the same way (so plai
```
Comment (trhr) [inline docs/source/evennia_core/system/portal/OOB.md:8]: no changes

File: docs/source/evennia_core/system/portal/Text-Encodings.md [hunk 21]
```diff
@@ -64,3 +64,8 @@ unneccesary overhead. Try to guess the most common encodings you players will
 use and make sure these are tried first. The International *UTF-8* encoding is
```
Comment (trhr) [inline docs/source/evennia_core/system/portal/Text-Encodings.md:64]: no changes

File: docs/source/evennia_core/system/portal/portal-server-architecture.md [hunk 22]
```diff
@@ -2,7 +2,7 @@
```
Comment (trhr) [inline docs/source/evennia_core/system/portal/portal-server-architecture.md:2]: no changes

File: docs/source/index.md [hunk 23]
```diff
@@ -1,31 +1,53 @@
 # Evennia Documentation
```
Comment (trhr) [inline docs/source/index.md:1]: rewrite

File: docs/source/related_topics/evennia-quickstart.md [hunk 24]
```diff
@@ -0,0 +1,24 @@
+```python
```
Comment (trhr) [inline docs/source/related_topics/evennia-quickstart.md:1]: restruc

File: docs/source/related_topics/technical/client-support-grid.md [hunk 25]
```diff
@@ -0,0 +1,86 @@
+```python
```
Comment (trhr) [inline docs/source/related_topics/technical/client-support-grid.md:1]: fixed a bunch of inaccuracies in the support grid. reorg'd clients. added column for OS.

File: docs/source/related_topics/technical/hosting-overview.md [hunk 26]
```diff
@@ -0,0 +1,208 @@
+```python
```
Comment (trhr) [inline docs/source/related_topics/technical/hosting-overview.md:1]: a lot of this was chopped out of other files. the point is that people who are 'getting started' don't need to know how this stuff in this level of detail, but people who are serverops for a production server do.

File: docs/source/related_topics/troubleshooting/linux-troubleshooting.md [hunk 27]
```diff
@@ -0,0 +1,25 @@
+```python
```
Comment (trhr) [inline docs/source/related_topics/troubleshooting/linux-troubleshooting.md:1]: a refactor of the troubleshooting page

File: docs/source/related_topics/troubleshooting/troubleshooting.md [hunk 28]
```diff
@@ -1,3 +1,45 @@
+```python
```
Comment (trhr) [inline docs/source/related_topics/troubleshooting/troubleshooting.md:1]: install troubleshooting -> this
quirks -> this
os-specific stuff -> individual pages

Comment (trhr) [reply docs/source/related_topics/troubleshooting/troubleshooting.md:1]: virtualenv stuff changed

File: docs/source/tutorials_and_examples/admin/Banning.md [hunk 29]
```diff
@@ -1,3 +1,7 @@
+```python
```
Comment (trhr) [inline docs/source/tutorials_and_examples/admin/Banning.md:1]: no changes

File: docs/source/tutorials_and_examples/admin/admin-overview.md [hunk 30]
```diff
@@ -0,0 +1,18 @@
+# Evennia Administration
```
Comment (trhr) [inline docs/source/tutorials_and_examples/admin/admin-overview.md:1]: WIP

File: docs/source/tutorials_and_examples/building/Builder-Docs.md [hunk 31]
```diff
@@ -0,0 +1,26 @@
+# Builder Docs
```
Comment (trhr) [inline docs/source/tutorials_and_examples/building/Builder-Docs.md:1]: WIP

File: docs/source/tutorials_and_examples/building/building-overview.md [hunk 32]
```diff
@@ -2,7 +2,7 @@
```
Comment (trhr) [inline docs/source/tutorials_and_examples/building/building-overview.md:2]: this is the 'quickstart' - no changes

File: docs/source/tutorials_and_examples/developing/development-overview.md [hunk 33]
```diff
@@ -0,0 +1,60 @@
+# Developer Tutorials
```
Comment (trhr) [inline docs/source/tutorials_and_examples/developing/development-overview.md:1]: WIP

File: docs/source/tutorials_and_examples/developing/utilities/Coding-Utils.md [hunk 34]
```diff
@@ -22,13 +22,13 @@ If you need to search for objects in a code module you can use the functions in
      obj = search_object(objname)
```
Comment (trhr) [inline docs/source/tutorials_and_examples/developing/utilities/Coding-Utils.md:22]: WIP - needs API links

File: docs/source/tutorials_and_examples/index.md [hunk 35]
```diff
@@ -0,0 +1,110 @@
+# Tutorials
```
Comment (trhr) [inline docs/source/tutorials_and_examples/index.md:1]: renamed from Tutorials.md; WIP needs restructure possibly

File: docs/source/tutorials_and_examples/python/virtualenv.md [hunk 36]
```diff
@@ -0,0 +1,31 @@
+```python
```
Comment (trhr) [inline docs/source/tutorials_and_examples/python/virtualenv.md:1]: substantial changes, needs review

File: docs/source/tutorials_and_examples/python/virtualenv.md [hunk 37]
```diff
@@ -0,0 +1,31 @@
...
+class Documentation:
+    RATING = "Acceptable" # Better structure
+```
+
+#Virtualenv
```
Comment (Griatch) [inline docs/source/tutorials_and_examples/python/virtualenv.md:6]: Needs a space between # and header, most likely?

File: docs/source/tutorials_and_examples/python/virtualenv.md [hunk 38]
```diff
@@ -0,0 +1,31 @@
...
+ 
+- `python -m virtualenv <name>` - initialize a new virtualenv `<name>` in a new folder `<name>` in the current location. Called `evenv` in these docs.
+- `python -m virtualenv -p path/to/alternate/python_executable <name>` - create a virtualenv using another Python version than default.
```
Comment (Griatch) [inline docs/source/tutorials_and_examples/python/virtualenv.md:9]: May be worth to offer an example as  `-p /usr/bin/python3.7` here. Overall, it can be a good idea to have the user check which is their default version is with `python --version` first.

File: docs/source/tutorials_and_examples/python/virtualenv.md [hunk 39]
```diff
@@ -0,0 +1,31 @@
...
+
+When your virtualenv is active you should see your console/terminal prompt change to 
+
+    (evenv) ...
+
+... or whatever name you gave the virtualenv when you initialized it. 
+
+> We sometimes say that we are "in" the virtualenv when it's active. But just to be clear - you never have to actually `cd` into the `evenv` folder. You can activate it from anywhere and will still be considered "in" the virtualenv wherever you go until you `deactivate` or close the console/terminal. 
+
+So, when do I *need* to activate my virtualenv? If the virtualenv is not active, none of the Python packages/programs you installed in it will be available to you. So at a minimum, *it needs to be activated whenever you want to use the `evennia` command* for any reason.
```
Comment (Griatch) [inline docs/source/tutorials_and_examples/python/virtualenv.md:30]: For consistency, it's probably best to stick to "you" rather than switching to "I".

File: evennia/accounts/migrations/0001_initial.py [hunk 40]
```diff
@@ -8,7 +8,7 @@
 
 class Migration(migrations.Migration):
 
-    dependencies = [("auth", "0001_initial"), ("typeclasses", "0001_initial")]
+    dependencies = [("auth", "0001_initial"), ("typeclasses", "0001_initial"), ("contenttypes", "__first__")]
```
Comment (Griatch) [inline evennia/accounts/migrations/0001_initial.py:11]: Hm, these definitely don't belong with this PR ...

File: docs/source/tutorials_and_examples/events/A-voice-operated-elevator-using-events.md [hunk 41]
```diff
@@ -1,8 +1,5 @@
 # A voice operated elevator using events
 
-
-- Previous tutorial: [Adding dialogues in events](Dialogues-in-events)
```
Comment (Griatch) [inline docs/source/tutorials_and_examples/events/A-voice-operated-elevator-using-events.md:4]: All this events-related stuff (this and the dialogues) is part of the In-Game-Python contrib, so it belongs under Contribs.

File: docs/source/community/documentation/DocumentationStyle.md [hunk 42]
```diff
@@ -0,0 +1,129 @@
...
+
+### Rating
+
+All pages receive a **rating** in the documentation. 
+
+```python
+RATING = "Exceptional"
+```
+
+The following ratings are available.
```
[the remark this answers is not in the document]
Comment (trhr) [reply docs/source/community/documentation/DocumentationStyle.md:88]: Ah, you're missing the key communication (which really hasn't been communicated well yet).

Exceptional-quality documentation is exceptional because the community _makes_ it exceptional.

It's designed to be the must-reads, that makes it the most-read. That makes it the most edited. That means if you see something wrong with it, definitely fix it. But maybe don't add a long paragraph about some random use case that you found; if you do, it may not get merged.

On the other hand, Needs Improvement-quality documentation is literally marked as such because we can expect to accept pretty much any changes to the documentation that are reasonable. If you think you want to put a completely different spin on it, and it's Needs Improvement, that WOULD probably get merged.

It's really a classification of how much scrutiny your changes will get, rather than a classification of whether or not you should try.

It also means that you can keep a list of a half-dozen must-read "exceptional" level documents that you regularly check. You know, like how your "installing evennia" still used python2 virtualenv, even though we've been off python2 for a long time? Feels like if that was one of the pieces of documentation that was regularly reviewed, that would have been caught right away.

File: evennia/accounts/migrations/0001_initial.py (hunk 40 above)
Comment (trhr) [reply evennia/accounts/migrations/0001_initial.py:11]: truth. they probably dont even belong in my directory.

File: docs/source/community/contributing/index.md [hunk 43]
```diff
@@ -0,0 +1,85 @@
+```python
+class Documentation:
+    RATING = "Excellent"
```
[the remark this answers is not in the document]
Comment (trhr) [reply docs/source/community/contributing/index.md:3]: i dumped everything in related_topics that I wasn't sure where it should live. I think some of it should move to tutorials and some of it should be cut from the docs entirely. i am not sure which is which; and think a fine tooth comb approach is required to refactoring this folder.

[the remark this answers is not in the document]
Comment (trhr) [reply docs/source/community/contributing/index.md:3]: a tutorials_and_examples/getting-started/ folder is definitely worth dumping a lot of the newbie documentation into.

File: docs/source/community/documentation/DocumentationStyle.md (hunk 42 above)
[the remark this answers is not in the document]
Comment (Griatch) [reply docs/source/community/documentation/DocumentationStyle.md:88]: I understand how you _intend_ the rating to be perceived, it's not how it will _be_ perceived. Most people will not be reading the instructions for what these ratings mean. And even so, it still comes across as very self-grandizing and arrogant to refer to itself as "Rating=Excellent" with no further qualifier. It just gives the wrong connotations. And yes, people will get discouraged from contributing by it. 

That said, it's not a bad idea to have something to track the important parts of the documentation. So maybe a rating `IMPORTANCE` or maybe `MAINTENANCE-IMPORTANCE` or something like that, with a rating meant for documentation-maintainers to be able to keep track of which documents really needs attention more regularly than others? 

### MAINTENANCE-IMPORTANCE

- Low
- Medium
- High
- Critical

### COMPLETENESS 

- Unknown
- Needs Improvements
- Complete

### COMPLEXITY 

- Beginner
- Intermediary
- Advanced 

> Btw: I still daily use virtualenv the same way as described in the installing evennia page; it works fine for python3, so changing that or not had nothing to do with with not giving the page attention ... ahem.

File: docs/source/community/contributing/index.md (hunk 43 above)
[the remark this answers is not in the document]
Comment (Griatch) [reply docs/source/community/contributing/index.md:3]: There are some lower-quality pages that could likely be cut, but most of this documentation _has_ been created after direct request by users. People learn in very different ways and some times it was easier to add to the documentation than having to answer the same question over and over. This is why we have detailed many things that are often not strictly evennia-specific; people ask us for the answer if they can't find it in our docs.

[the remark this answers is not in the document]
Comment (Griatch) [reply docs/source/community/contributing/index.md:3]: And again, using sub-subfolders seems counter-productive overall - the more nested this structure is, the harder it is to get an overview of what we actually have, with the risk increasing that some docs are completely forgotten about. Better to have a flatter structure than that, as said.

**Comment by Griatch** [comment]
@trhr Are you planning to address or comment on the feedback I gave in the last review or is the PR meant to stay as a sort of reference for now?

**Comment by Griatch** [comment]
(As discussed, this will be not be merged at this time; but may be something used for a future new version of the docs.)

**Comment by Griatch** [comment]
@trhr: I have now completed my work around the new documentation infrastructure, which includes all the build mechanisms and auto-documentation but also the move of the documentation sources into a preliminary structure. The server docs are grouped only within the very coarse categories Setup, Coding, Components, Concepts and Howtos. The whole thing is planned to be built around a central multi-part starting tutorial where many of the previous howto's and tutorials are gathered with a much clearer red thread (along with a lot of new material not yet created). 

There are still a lot to do on that, but it's clear that **this PR will not be merging**. I don't want to lose the work you already spent on reworking individual doc-pages though; so I'll keep this open for now to see what is still relevant.
